src/qibo/parameter.py

```diff
@@ -0,0 +1,109 @@
+    final gate parameter. All possible analytical derivatives of the lambda function are
+    calculated at the object initialisation using Sympy.
+
+    Example::
+
+        from qibo.parameter import Parameter
+        param = Parameter(
+                lambda x, th1, th2, th3: x**2 * th1 + th2 * th3**2,
+                [1.5, 2.0, 3.0],
+                feature=[7.0],
+            )
+
+        partial_derivative = param.get_partial_derivative(3)
+
+        param.update_parameters(trainable=[15.0, 10.0, 7.0], feature=[5.0])
+        gate_value = param()
+
+
+    Args:
+        func (function): lambda function which builds the gate parameter. If both features and trainable parameters
+        compose the function, it must be passed by first providing the features and then the parameters, as
+        described in the code example above.
+        trainable (list or np.ndarray): array with initial trainable parameters theta
+        feature (list or np.ndarray): array containing possible input features x
+    """
+
+    def __init__(self, func, trainable, feature=None):
+        self._trainable = trainable
+        self._feature = feature
+
+        # lambda function
+        self.lambdaf = func
+
+        self.derivatives = self._calculate_derivatives()
+
+    def __call__(self):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        return self._apply_func(self.lambdaf)
+
+    @property
+    def nparams(self):
+        return len(self._trainable)
+
+    @property
+    def nfeat(self):
+        return len(self._feature) if isinstance(self._feature, list) else 0
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._feature is not None:
+            params.extend(self._feature)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainable)
+
+        # run function
+        return float(function(*params))
+
+    def _calculate_derivatives(self):
+        """Calculates derivatives w.r.t to all trainable parameters"""
+        vars = []
+        for i in range(self.nfeat):
+            vars.append(sp.Symbol(f"x{i}"))
+        for i in range(self.nparams):
+            vars.append(sp.Symbol(f"th{i}"))
+
+        expr = sp.sympify(self.lambdaf(*vars))
+
+        derivatives = []
+        for i in range(len(vars)):
+            derivative_expr = sp.diff(expr, vars[i])
+            derivatives.append(sp.lambdify(vars, derivative_expr))
+
+        return derivatives
+
+    def update_parameters(self, trainable=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if trainable is not None:
+            self._trainable = trainable
+
+        if feature is not None and self._feature is not None:
+            self._feature = feature
```

Comment:
I have the impression to have already written it, but: why not using just property setters?

https://docs.python.org/3/library/functions.html#property

Reply:
Even better: since in this case the setters would be trivial, why not just exposing them as `self.trainable` and `self.feature`?

If they could be directly modified by the user, no need to hide them (and no need for this method).

Reply:
Yes, maybe to remove the method is the cleaner way. One can modify the features by directly calling `self.features = new_features`.

Reply:
> Overall, I think it is not very clean to change object variables directly, as the user might not know what format they take. Usually it is encouraged to interact with objects through functions instead.

This is exactly the Java perspective. But Python has `@property`s exactly for that, while providing a more ergonomic syntax.
But if you are not doing anything in your function, other than setting the parameter, there is no added benefit from the function itself (other than the syntax and call overhead...).

However, if you want to accept `None` and ignore it (rather than literally setting to `None`), then the setter is non-trivial, you're right. Thus, I would propose the `@property` approach, and just split `trainable` and `feature` in two separate properties.